cpp/src/arrow/ipc/writer.cc

```diff
@@ -329,15 +329,24 @@ class RecordBatchSerializer {
       return Status::OK();
     }
 
-    int64_t required_bytes = sizeof(offset_type) * (array.length() + 1);
-    if (array.value_offset(0) > 0) {
+    const int64_t required_bytes = sizeof(offset_type) * (array.length() + 1);
+
+    offset_type first_offset = 0;
+    RETURN_NOT_OK(MemoryManager::CopyBufferSliceToCPU(
+        array.data()->buffers[1], array.offset() * sizeof(offset_type),
+        sizeof(offset_type), reinterpret_cast<uint8_t*>(&first_offset)));
+
+    if (first_offset > 0) {
       // If the offset of the first value is non-zero, then we must create a new
       // offsets buffer with shifted offsets.
+      if (!array.data()->buffers[1]->is_cpu()) {
+        return Status::NotImplemented("Rebasing non-CPU offsets");
+      }
```

Comment:
Why do we need this check?
May `MemoryManager::CopyBufferSliceToCPU()` return a memory not on CPU?

Reply:
Because we copied a single offset above and we're accessing the entire buffer here.

Reply:
Ah, I misunderstood `MemoryManager::CopyBufferSliceToCPU()`.
I thought that `MemoryManager::CopyBufferSliceToCPU()` updates `array.data()->buffers[1]`. But it copies to `first_offset`.